File: Source/NextTurnRuntime/Private/CoreUObject/WeakObjectReference.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Private/Accessor.hpp"
#include "NextTurnRuntime/Private/Construct.hpp"
#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/CoreUObject/Public/UObject/WeakObjectPtr.h"

ENABLE_WARNINGS;

struct FNextTurnWeakObjectReference
{
	int32_t Index;
	int32_t SerialNumber;
};

DEFINE_SIZE_ACCESSOR(FWeakObjectPtr, ObjectIndex);
DEFINE_SIZE_ACCESSOR(FWeakObjectPtr, ObjectSerialNumber);

static_assert(sizeof(FNextTurnWeakObjectReference) == sizeof(FWeakObjectPtr), "");

static_assert(sizeof(FNextTurnWeakObjectReference::Index) == SIZEOF(FWeakObjectPtr, ObjectIndex), "");
static_assert(sizeof(FNextTurnWeakObjectReference::SerialNumber) == SIZEOF(FWeakObjectPtr, ObjectSerialNumber), "");

#define TYPE_EXPORT_MANAGED_TYPE_NAME WeakObjectReference

TYPE_EXPORT_START

void EXPORT_CALL_CONV Initialize(FWeakObjectPtr& Reference, const UObject* Target)
{
	Construct(Reference, Target);
}

UObject* EXPORT_CALL_CONV GetTarget(const FWeakObjectPtr& Reference)
{
	return Reference.Get();
}

bool EXPORT_CALL_CONV IsValid(const FWeakObjectPtr& Reference)
{
	return Reference.IsValid();
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(Initialize),
	EXPORT_METHOD(GetTarget),
	EXPORT_METHOD(IsValid),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

File: Managed/NextTurn.UE.Runtime/CoreUObject/WeakObjectReference.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using System.Diagnostics.CodeAnalysis;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public readonly struct WeakObjectReference : IObjectReference, IEquatable<WeakObjectReference>
    {
        private readonly int index;
        private readonly int serialNumber;

        public WeakObjectReference(Object? target) => NativeMethods.Initialize(out this, Object.GetPointerOrZero(target));

        public bool IsValid => NativeMethods.IsValid(this);

        public Object? Target => Object.CreateOrNull<Object>(NativeMethods.GetTarget(this));

        public override bool Equals(object? value) => value is WeakObjectReference other && this.Equals(other);

        public bool Equals(WeakObjectReference other) =>
            (this.index == other.index && this.serialNumber == other.serialNumber) ||
            (!this.IsValid && !other.IsValid);

        public override int GetHashCode() => this.index ^ this.serialNumber;

        public bool TryGetTarget([NotNullWhen(true)] out Object? target) => (target = this.Target) != null;

        public static bool operator ==(WeakObjectReference left, WeakObjectReference right) => left.Equals(right);

        public static bool operator !=(WeakObjectReference left, WeakObjectReference right) => !(left == right);

        internal static class NativeMethods
        {
            [Calli]
            public static extern void Initialize(out WeakObjectReference reference, IntPtr target);

            [Calli]
            public static extern IntPtr GetTarget(in WeakObjectReference reference);

            [Calli]
            public static extern bool IsValid(in WeakObjectReference reference);
        }
    }
}
